File: app/internal/publisher/template/go/widgets/dashboard_widget.go
package widgets

import (
	"fmt"

	"github.com/websmith/sdk"
)

type Dashboard struct {
	engine    *wsd.LicenseEngine
	status    string
	daysLeft  int
	expiry    string
	plan      string
}

func NewDashboard(engine *wsd.LicenseEngine) *Dashboard {
	return &Dashboard{
		engine: engine,
	}
}

func (d *Dashboard) Render() string {
	d.Refresh()
	output := ""
	output += fmt.Sprintf("License Status: %s\n", d.status)
	if d.daysLeft > 0 {
		output += fmt.Sprintf("Days Remaining: %d\n", d.daysLeft)
	}
	if d.expiry != "" {
		output += fmt.Sprintf("Expiry: %s\n", d.expiry)
	}
	if d.plan != "" {
		output += fmt.Sprintf("Plan: %s\n", d.plan)
	}
	return output
}

func (d *Dashboard) Refresh() {
	s := d.engine.GetStatus()
	if s == nil {
		s = d.engine.Initialize()
	}
	if s != nil && s.Valid && (s.Status == "active" || s.Status == "trial" || s.Status == "trial_active") {
		d.status = "Active"
		if s.TrialActive {
			d.status = "Trial Active"
		}
		d.daysLeft = s.DaysRemaining
		d.expiry = s.ExpiresAt
		d.plan = s.Plan
	} else {
		d.status = "Unlicensed"
		d.daysLeft = 0
		d.expiry = ""
		d.plan = ""
	}
}

func (d *Dashboard) GetStatus() string {
	return d.status
}

func (d *Dashboard) GetDaysLeft() int {
	return d.daysLeft
}

func (d *Dashboard) GetExpiry() string {
	return d.expiry
}

func (d *Dashboard) GetPlan() string {
	return d.plan
}

File: app/internal/publisher/template/rust/src/widgets/dashboard_widget.rs
use crate::license_engine::LicenseEngine;

pub struct DashboardWidget {
    engine: *mut LicenseEngine,
    status: String,
    days_left: u32,
    expiry: String,
    plan: String,
}

impl DashboardWidget {
    pub fn new(engine: &mut LicenseEngine) -> Self {
        Self {
            engine: engine as *mut LicenseEngine,
            status: String::new(),
            days_left: 0,
            expiry: String::new(),
            plan: String::new(),
        }
    }

    pub fn render(&mut self) -> String {
        self.refresh();
        let mut output = String::new();
        output.push_str(&format!("License Status: {}\n", self.status));
        if self.days_left > 0 {
            output.push_str(&format!("Days Remaining: {}\n", self.days_left));
        }
        if !self.expiry.is_empty() {
            output.push_str(&format!("Expiry: {}\n", self.expiry));
        }
        if !self.plan.is_empty() {
            output.push_str(&format!("Plan: {}\n", self.plan));
        }
        output
    }

    pub fn refresh(&mut self) {
        let engine = unsafe { &mut *self.engine };
        let s = engine.get_status_cloned().unwrap_or_else(|| engine.initialize());
        if s.valid && (s.status == "active" || s.status == "trial" || s.status == "trial_active") {
            self.status = if s.trial_active {
                "Trial Active".to_string()
            } else {
                "Active".to_string()
            };
            self.days_left = s.days_remaining;
            self.expiry = s.expires_at.unwrap_or_default();
            self.plan = s.plan.unwrap_or_default();
        } else {
            self.status = "Unlicensed".to_string();
            self.days_left = 0;
            self.expiry.clear();
            self.plan.clear();
        }
    }

    pub fn get_status(&self) -> &str {
        &self.status
    }

    pub fn get_days_left(&self) -> u32 {
        self.days_left
    }
}
